### ISSUE TITLE
accuracy not high

### ISSUE THREAD

**Issue by badcode-iip** [bdsaglam/torch-scae#5]
Hello sir
Nice work! However, I trained with the hparams as you suggested, and it shows that the validation accuracy is about 0.7, while the paper reported their accuracy is 98.7%. So I was wondering if I did something wrong or you got the same result?

**Comment by bdsaglam** [comment]
Hi, I haven't trained it for long as the original study. However, yesterday I fixed a bug that improved convergence behavior. Can you install version 0.3 and train it again? Also, in the original study, image augmentation is employed but it is missing in this repo, so it might be another reason for the difference. I'll will implement image augmentation soon.

**Comment by badcode-iip** [comment]
Thanks for your reply.
version 0.3 performs better. The model converges after about 100 epochs, validation accuracy reaches 0.85. Looking forward to your follow-up work.

**Comment by bdsaglam** [comment]
Cool, thanks for sharing it. I have added image augmentation for MNIST as in the original study. It pads each image to size of 40 and randomly shifts in vertical and horizontal direction. You can try it now.

**Comment by badcode-iip** [comment]
Hello sir
It seems that something is wrong about reconstruction likelihood? The rec_ll_loss stays high during the whole training.

**Comment by badcode-iip** [comment]
Hello sir
It seems that something is wrong about reconstruction likelihood? The rec_ll_loss stays high during the whole training progress. Also the rec_ll_loss is much bigger than the other loss component. Did I get it wrong?

**Comment by bdsaglam** [comment]
Hi, 
With default hyperparameters, what I observed is that reconstruction loss decreases very slowly. But, it seems same with the original implementation. Somehow, it takes model many steps to learn templates. How long did you train it and with what hyperparameters?

**Comment by badcode-iip** [comment]
Hi,
I tried the default hyperparameters, which is same as the tensorflow version (batchsize=128, learning rate=3e-5) .But the result is worse than before(batchsize=32, learning rate=1e-4). 
I alse tried tensorflow version, and it seems that the loss curves are different.

tensorflow version:

![图片](https://user-images.githubusercontent.com/25474159/84646176-45a7b380-af34-11ea-971e-90a23e8e9129.png)

![图片](https://user-images.githubusercontent.com/25474159/84646207-51937580-af34-11ea-82ca-f248aef25b02.png)

![图片](https://user-images.githubusercontent.com/25474159/84646484-c1a1fb80-af34-11ea-8104-82f3c103f2c9.png)


pytorch lightning version:
![图片](https://user-images.githubusercontent.com/25474159/84645855-d4680080-af33-11ea-9c53-f06c6126e43b.png)

![图片](https://user-images.githubusercontent.com/25474159/84645562-5a377c00-af33-11ea-9265-d380f82f0cb0.png)

![图片](https://user-images.githubusercontent.com/25474159/84646526-cff01780-af34-11ea-89c6-9a8b23117019.png)

**Comment by bdsaglam** [comment]
Hi, thanks for sharing those. The result might be worse than previous version because it augments images on default. Also, I noticed that I did not turn learning rate scheduler on default, which is used in the original implementation. Changed it now. But, this is probably not the root cause. There might be a bug somewhere. I will check it when I have time.

**Comment by cxiang26** [comment]
I got the low accuracy as well, which configs could get more precise?
![image](https://user-images.githubusercontent.com/22794673/96061492-f19ee280-0ec5-11eb-95a5-57f49734e5f5.png)

**Comment by cxiang26** [comment]
> Thanks for your reply.
> version 0.3 performs better. The model converges after about 100 epochs, validation accuracy reaches 0.85. Looking forward to your follow-up work.

Hi guys, could you tell me where to find version 0.3?
Thanks in advance.

**Comment by bdsaglam** [comment]
> > Thanks for your reply.
> > version 0.3 performs better. The model converges after about 100 epochs, validation accuracy reaches 0.85. Looking forward to your follow-up work.
> 
> Hi guys, could you tell me where to find version 0.3?
> Thanks in advance.

Hi cxiang26, 
I did not create releases in this project, sorry for my inexperience. I guess you can check out to the commit `cefabf93b8d65fb21cd14e3ff57efb473e055231` to get version 0.3.

**Comment by cxiang26** [comment]
We get a stable acc. over 80% if using grad_clip=2 or lesser.
Hope this is helpful.

![image](https://user-images.githubusercontent.com/22794673/98192169-7fb23a00-1f55-11eb-9a0c-4f976b49464e.png)
![image](https://user-images.githubusercontent.com/22794673/98192194-8f318300-1f55-11eb-84e1-f9fdeef7601e.png)

**Comment by cxiang26** [comment]
> We get a stable acc. over 80% if using grad_clip=2 or lesser.
> Hope this is helpful.
> 
> ![image](https://user-images.githubusercontent.com/22794673/98192169-7fb23a00-1f55-11eb-9a0c-4f976b49464e.png)
> ![image](https://user-images.githubusercontent.com/22794673/98192194-8f318300-1f55-11eb-84e1-f9fdeef7601e.png)

**Comment by cxiang26** [comment]
@bdsaglam 
Hi guys, I find this code could have below bugs compared to reimplemented experiments by myself.
BUG1: I think that the key bug of this code is the initialization problem for the templates. I got excellent results if I set the value of initialization templates to [0, 0.1] rather than [0, 1]. solution: adding a line code `q = 0.1 * q` after line 67  of part_decoder.py.
BUG2: Don't use Relu activation in the final layer of the caps_mlps network, which could lead to the value of cpr_dynamic larger than 0 always,  that makes the training is unstable prettily. Solution:  set 'activate_final=False' in object_decoder.py line 105.
![image](https://user-images.githubusercontent.com/22794673/99074171-38c9e180-25f2-11eb-8a33-bd4084e31816.png)

**Comment by liuweijie19980216** [comment]
> @bdsaglam
> Hi guys, I find this code could have below bugs compared to reimplemented experiments by myself.
> BUG1: I think that the key bug of this code is the initialization problem for the templates. I got excellent results if I set the value of initialization templates to [0, 0.1] rather than [0, 1]. solution: adding a line code `q = 0.1 * q` after line 67 of part_decoder.py.
> BUG2: Don't use Relu activation in the final layer of the caps_mlps network, which could lead to the value of cpr_dynamic larger than 0 always, that makes the training is unstable prettily. Solution: set 'activate_final=False' in object_decoder.py line 105.
> ![image](https://user-images.githubusercontent.com/22794673/99074171-38c9e180-25f2-11eb-8a33-bd4084e31816.png)



Hello, I run version 0.3, the accuracy has been around 0.1, can you share your version？？？

**Comment by tuwangwenx** [comment]
> > We get a stable acc. over 80% if using grad_clip=2 or lesser.
> > Hope this is helpful.
> > ![image](https://user-images.githubusercontent.com/22794673/98192169-7fb23a00-1f55-11eb-9a0c-4f976b49464e.png)
> > ![image](https://user-images.githubusercontent.com/22794673/98192194-8f318300-1f55-11eb-84e1-f9fdeef7601e.png)



> @bdsaglam Hi guys, I find this code could have below bugs compared to reimplemented experiments by myself. BUG1: I think that the key bug of this code is the initialization problem for the templates. I got excellent results if I set the value of initialization templates to [0, 0.1] rather than [0, 1]. solution: adding a line code `q = 0.1 * q` after line 67 of part_decoder.py. BUG2: Don't use Relu activation in the final layer of the caps_mlps network, which could lead to the value of cpr_dynamic larger than 0 always, that makes the training is unstable prettily. Solution: set 'activate_final=False' in object_decoder.py line 105. ![image](https://user-images.githubusercontent.com/22794673/99074171-38c9e180-25f2-11eb-8a33-bd4084e31816.png)

你好，请问一下你这个图片上的结果怎么得到的啊，我试了你说的方法，但是正确率还是在60%多，可不可以分享一下你得出这个实验结果的代码啊，求求了。

**Comment by tuwangwenx** [comment]
> > We get a stable acc. over 80% if using grad_clip=2 or lesser.
> > Hope this is helpful.
> > ![image](https://user-images.githubusercontent.com/22794673/98192169-7fb23a00-1f55-11eb-9a0c-4f976b49464e.png)
> > ![image](https://user-images.githubusercontent.com/22794673/98192194-8f318300-1f55-11eb-84e1-f9fdeef7601e.png)

@cxiang26 你好，请问一下你这个图片上的结果怎么得到的啊，我试了你说的方法，但是正确率还是在60%多，可不可以分享一下你得出这个实验结果的代码啊，求求了。

**Comment by cxiang26** [comment]
@tuwangwenx 很久没有弄这个了， 你试试我的项目，应该可以的 https://github.com/cxiang26/stacked_capsule_autoencode_pl

**Comment by liujiaxing1204** [comment]
> Thanks for your reply. version 0.3 performs better. The model converges after about 100 epochs, validation accuracy reaches 0.85. Looking forward to your follow-up work.

could you share the version you have traied?  I tried many versions and only got the val_acc of about 13%.

**Comment by bdsaglam** [comment]
Fixed in #16. The main bug was that part templates were warped with the pose instead of its inverse, which blew every template up to at least the size of the canvas. With that and a few smaller mismatches fixed, 300k steps on 40x40 MNIST give 98.6–98.7% k-means test accuracy (paper: 98.7%). See the README for results and `torch_scae_experiments/mnist/evaluate.py` for the evaluation. Thanks to everyone who dug into this.
